pkg/cli/preview/result.go

```diff
@@ -20,74 +20,182 @@ import (
+		directBad = fmt.Sprintf("%d", total-*s.directCount)
+	}
+	dcl := "NA"
+	dclBad := "NA"
+	if s.dclCount != nil {
+		dcl = fmt.Sprintf("%d", *s.dclCount)
+		dclBad = fmt.Sprintf("%d", total-*s.dclCount)
+	}
+	custom := "NA"
+	customBad := "NA"
+	if s.customCount != nil {
+		custom = fmt.Sprintf("%d", *s.customCount)
+		customBad = fmt.Sprintf("%d", total-*s.customCount)
+	}
+	return fmt.Sprintf("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s", tf, tfBad, direct, directBad, dcl, dclBad, custom, customBad)
+}
+
+type GKN struct {
+	Group     string
+	Kind      string
+	Namespace string
+}
+
 type PreviewSummary struct {
-	totalGood int
-	totalBad  int
-	// namespace -> group -> kind -> []name
-	goodGKNN map[string]map[string]map[string][]string
-	badGKNN  map[string]map[string]map[string][]string
-	reports  map[GKNN][]string
+	// namespace -> group -> kind -> GKNSummary
+	// Count the number of resources successfully reconciled by each reconciler type
+	summary map[string]map[string]map[string]*GKNSummary
+	// GKNN -> controller type -> []string
+	// reports store the GKNN and controller type that has issues
+	reports map[GKNN]map[k8s.ReconcilerType][]string
+	// namespace -> group -> kind -> count
+	// Count the total number of resources in each namespace, group, kind
+	gknCount map[string]map[string]map[string]int
 }
 
-func (r *Recorder) newPreviewSummary() *PreviewSummary {
-	summary := &PreviewSummary{
-		totalGood: 0,
-		totalBad:  0,
-		goodGKNN:  make(map[string]map[string]map[string][]string),
-		badGKNN:   make(map[string]map[string]map[string][]string),
-		reports:   make(map[GKNN][]string),
+// add result to the summary
+func (p *PreviewSummary) addResult(gknn GKNN, r ReconcileResult) {
+	if p.summary[gknn.Namespace] == nil {
+		p.summary[gknn.Namespace] = make(map[string]map[string]*GKNSummary)
+	}
+	if p.summary[gknn.Namespace][gknn.Group] == nil {
+		p.summary[gknn.Namespace][gknn.Group] = make(map[string]*GKNSummary)
+	}
+	summary := p.summary[gknn.Namespace][gknn.Group][gknn.Kind]
+	if summary == nil {
+		summary = &GKNSummary{}
+		p.summary[gknn.Namespace][gknn.Group][gknn.Kind] = summary
 	}
 
-	for gknn, info := range r.objects {
-		ensureGKNNPath(summary.goodGKNN, gknn)
-		ensureGKNNPath(summary.badGKNN, gknn)
-
-		good, report := ParseEventInfo(info)
-		if good {
-			summary.totalGood++
-			summary.goodGKNN[gknn.Namespace][gknn.Group][gknn.Kind] = append(summary.goodGKNN[gknn.Namespace][gknn.Group][gknn.Kind], gknn.Name)
-		} else {
-			summary.totalBad++
-			summary.badGKNN[gknn.Namespace][gknn.Group][gknn.Kind] = append(summary.badGKNN[gknn.Namespace][gknn.Group][gknn.Kind], gknn.Name)
-			summary.reports[gknn] = report
+	var countPtr **int
+	switch r.controllerType {
+	case k8s.ReconcilerTypeTerraform:
+		countPtr = &summary.tfCount
+	case k8s.ReconcilerTypeDCL:
+		countPtr = &summary.dclCount
+	case k8s.ReconcilerTypeDirect:
+		countPtr = &summary.directCount
+	default:
+		countPtr = &summary.customCount
+	}
+
+	if r.good {
```

Comment:
Suggestion for the future: `type PreviewSummary` has probably hit the limit of what we can sustainably maintain.  You may want to just try a totally different approach.  Splitting it into a type with a few methods and a private data representation is a good first step there.